Scripts/Data/Collection/ObservableCounter.cs

```diff
@@ -0,0 +1,114 @@
+﻿namespace VRTK.Core.Data.Collection
+{
+    using UnityEngine;
+    using UnityEngine.Events;
+    using System.Linq;
+    using System.Collections.Generic;
+
+    /// <summary>
+    /// Allows observing counting the amount of attempts an element is added or removed from a <see cref="Dictionary{TKey, TValue}"/>.
+    /// </summary>
+    /// <typeparam name="TElement">The type of the elements to count.</typeparam>
+    /// <typeparam name="TEvent">The <see cref="UnityEvent"/> type to use.</typeparam>
+    public class ObservableCounter<TElement, TEvent> : MonoBehaviour where TEvent : UnityEvent<TElement>, new()
+    {
+        /// <summary>
+        /// Emitted when a searched element is found.
+        /// </summary>
+        public TEvent ElementAdded = new TEvent();
+        /// <summary>
+        /// Emitted when a searched element is not found.
+        /// </summary>
+        public TEvent ElementRemoved = new TEvent();
+
+        /// <summary>
+        /// The elements being counted.
+        /// </summary>
+        public Dictionary<TElement, int> ElementsCounter
+        {
+            get;
+            protected set;
+        } = new Dictionary<TElement, int>();
+
+        /// <summary>
+        /// Increases the count on the given element.
+        /// </summary>
+        /// <param name="element">The element to count.</param>
+        public virtual void IncreaseCount(TElement element)
+        {
+            if (!isActiveAndEnabled || element == null)
+            {
+                return;
+            }
+
+            if (!ElementsCounter.ContainsKey(element))
+            {
+                ElementsCounter.Add(element, 0);
+            }
+
+            ElementsCounter[element]++;
+            if (ElementsCounter[element] == 1)
+            {
+                ElementAdded?.Invoke(element);
+            }
+        }
+
+        /// <summary>
+        /// Decreases the count on the given element.
+        /// </summary>
+        /// <param name="element">The element to count.</param>
+        public virtual void DecreaseCount(TElement element)
+        {
+            int currentValue = 0;
+            if (!isActiveAndEnabled || element == null || !ElementsCounter.TryGetValue(element, out currentValue) || currentValue <= 0)
+            {
+                return;
+            }
+
+            ElementsCounter[element]--;
+            if (ElementsCounter[element] <= 0)
+            {
+                ElementsCounter.Remove(element);
+                ElementRemoved?.Invoke(element);
+            }
+        }
+
+        /// <summary>
+        /// Removes the element from the counter.
+        /// </summary>
+        /// <param name="element">The element to clear.</param>
+        public virtual void RemoveFromCount(TElement element)
+        {
+            if (!isActiveAndEnabled || element == null || !ElementsCounter.ContainsKey(element))
```

Comment:
This dictionary lookup can be removed since `.Remove` actually returns a bool!
```suggestion
            if (!isActiveAndEnabled || element == null)
```

Reply:
I need to get the value out at that point to check if the value is less or equal zero to bomb out too.

If i do a remove there, then the following `ElementsCounter[element]--` will fail too

Reply:
This is the clear method which doesn't do any counter--. Are you confusing this method with the decrease one?

Reply:
I was 😄